src/TheNextWeek/planar.h

```diff
@@ -0,0 +1,158 @@
+#ifndef PLANAR_H
+#define PLANAR_H
+//==============================================================================================
+// To the extent possible under law, the author(s) have dedicated all copyright and related and
+// neighboring rights to this software to the public domain worldwide. This software is
+// distributed without any warranty.
+//
+// You should have received a copy (see file COPYING.txt) of the CC0 Public Domain Dedication
+// along with this software. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//==============================================================================================
+
+#include "rtweekend.h"
+
+#include "hittable.h"
+
+#include <cmath>
+
+//// This class is somewhat complex, but that's because it serves as a base class for very
+//// simple derived classes, where you really only need to worry about construction, bounds,
+//// and UV hit determination.
+
+class planar : public hittable {
+  public:
+    planar(const point3& o, const vec3& side_A, const vec3& side_B, shared_ptr<material> m)
+      : origin(o), a(side_A), b(side_B), mat(m)
+    {
+        normal = unit_vector(cross(a, b));
+        D = -dot(normal, origin);
+        a /= dot(a,a);
+        b /= dot(b,b);
+
+/////// By default, bounded to the world coordinates of UV<0,0> to UV<1,1>.
+
+        interval u_bounds, v_bounds;
+        get_uv_bounds(u_bounds, v_bounds);
+        bbox = aabb(
+            origin + (u_bounds.min * side_A) + (v_bounds.min * side_B),
+            origin + (u_bounds.max * side_A) + (v_bounds.max * side_B)
```

Comment:
You mentioned problems with having an infinite plane.
As is, this plane appears to be finite.

Reply:
Yes. This one is an abstract base class, so just tends to the common case.